Alc/effects/autowah.c

```diff
@@ -0,0 +1,266 @@
+static inline ALfloat envelope_follower(ALautowahState *state, ALfloat SampleIn)
+{
+    ALfloat alpha, Sample;
+
+    Sample =  state->PeakGain*fabsf(SampleIn);
+    alpha  = (Sample > state->env_delay) ? state->AttackRate : state->ReleaseRate;
+    state->env_delay = alpha*state->env_delay + (1.0f-alpha)*Sample;
+
+    return state->env_delay;
+}
+
+static void ALautowahState_Construct(ALautowahState *state)
+{
+    ALeffectState_Construct(STATIC_CAST(ALeffectState, state));
+    SET_VTABLE2(ALautowahState, ALeffectState, state);
+}
+
+static ALvoid ALautowahState_Destruct(ALautowahState *state)
+{
+    ALeffectState_Destruct(STATIC_CAST(ALeffectState,state));
+}
+
+static ALboolean ALautowahState_deviceUpdate(ALautowahState *state, ALCdevice *UNUSED(device))
+{
+    /* (Re-)initializing parameters and clear the buffers. */
+    state->AttackRate    = 1.0f;
+    state->ReleaseRate   = 1.0f;
+    state->ResonanceGain = 10.0f;
+    state->PeakGain      = 4.5f;
+    state->FreqMinNorm   = 4.5e-4f;
+    state->BandwidthNorm = 0.05f;
+    state->env_delay     = 0.0f;
+
+    BiquadFilter_clear(&state->Filter);
+
+    memset(state->CurrentGains, 0, sizeof(state->CurrentGains));
+    memset(state->TargetGains,  0, sizeof(state->TargetGains));
+
+    return AL_TRUE;
+}
+
+static ALvoid ALautowahState_update(ALautowahState *state, const ALCcontext *context, const ALeffectslot *slot, const ALeffectProps *props)
+{
+    const ALCdevice *device = context->Device;
+    ALfloat coeffs[MAX_AMBI_COEFFS];
+    ALfloat ReleaseTime;
+
+    ReleaseTime = clampf(props->Autowah.ReleaseTime,0.001f,1.0f);
+
+    state->AttackRate    = expf(-1.0f/(props->Autowah.AttackTime*device->Frequency));
+    state->ReleaseRate   = expf(-1.0f/(ReleaseTime*device->Frequency));
+    state->ResonanceGain = 10.0f/3.0f*log10f(props->Autowah.Resonance);/*0-20dB Resonance Peak gain*/
+    state->PeakGain      = 1.0f -log10f(props->Autowah.PeakGain/AL_AUTOWAH_MAX_PEAK_GAIN);
+    state->FreqMinNorm   = MIN_FREQ/device->Frequency;
+    state->BandwidthNorm = (MAX_FREQ - MIN_FREQ)/device->Frequency;
+
+    CalcAngleCoeffs(0.0f, 0.0f, 0.0f, coeffs);
+    ComputeDryPanGains(&device->Dry, coeffs, slot->Params.Gain, state->TargetGains);
+}
+
+static ALvoid ALautowahState_process(ALautowahState *state, ALsizei SamplesToDo, const ALfloat (*restrict SamplesIn)[BUFFERSIZE], ALfloat (*restrict SamplesOut)[BUFFERSIZE], ALsizei NumChannels)
+{
+    ALfloat *restrict BufferOut = state->BufferOut;
+    ALsizei i;
+
+    for(i = 0;i < SamplesToDo;i++)
+    {
+        ALfloat env_out, f0norm, temp;
+
+        env_out = envelope_follower(state, SamplesIn[0][i]);
+        f0norm = state->BandwidthNorm*env_out + state->FreqMinNorm;
+
+        BiquadFilter_setParams(&state->Filter, BiquadType_Peaking, state->ResonanceGain,
+                               f0norm, 1.0f/Q_FACTOR);
+        BiquadFilter_process(&state->Filter, &temp, &SamplesIn[0][i], 1);
```

Comment:
Unfortunately, this could be trouble. Processing 1 sample at a time ist going to be pretty inefficient. Plus, I don't think the current biquad processing loop is very robust to the coefficients constantly changing. It might be better to process a group of samples at a time (64, 128?), using a look-ahead buffer/delay to ensure the calculated envelope aligns with the output, so as to not overwhelm the filter.

Reply:
I have made a stage of adjustment and optimization of a month trying for several hours the filter and doesn't seem to have had problems. Only when AttackTime and ReleaseTime were equal to 0.0001 I had stability problems, so I had to do a clamp to ReleaseTime to 0.001.
`ReleaseTime = clampf(props->Autowah.ReleaseTime,0.001f,1.0f);`

>I don't think the current biquad processing loop is very robust to the coefficients constantly changing. It might be better to process a group of samples at a time (64, 128?), using a look-ahead buffer/delay to ensure the calculated envelope aligns with the output, so as to not overwhelm the filter.

You may be right, although you probably have to change more parameters to have the same behavior that it currently has. I'll take a look in the next weeks if I have spare time, but I can tell you that the filter is stable.

Reply:
Any optimization is always good, the important thing is that the algorithm works and we can offer functionality to this effect in OpenAL soft now.